hammers/scripts/unutilized_lease_reaper.py

```diff
@@ -0,0 +1,178 @@
+        for k, v in blazar.hosts(auth).items()}
+    nodes_by_host = {
+        hosts_by_node_uuid[k]: v for k, v
+        in ironic.nodes(auth, details=True).items()}
+    allocations = blazar.host_allocations(auth)
+    allocs_by_lease = defaultdict(list)
+
+    for alloc in allocations:
+        for reservation in alloc['reservations']:
+            allocs_by_lease[reservation['lease_id']].append(
+                nodes_by_host[alloc['resource_id']])
+
+    for lease in leases:
+        lease['nodes'] = allocs_by_lease[lease['id']]
+
+    return leases
+
+
+def send_notification(lease, sender, subject, email_body):
+    user = keystone.user(lease['user_id'])
+    html = _email.render_template(
+        email_body,
+        vars=dict(lease_name=lease['name'],
+                  lease_id=lease['id']))
+    _email.send(_email.get_host(), user['email'], sender, subject,
+                html.encode('utf8'))
+
+
+def reaper(auth, sender, warn_period, termination_period, delete=False):
+    leases = leases_with_node_details(auth)
+    leases_to_warn = filter(inviolation_filter(warn_period), leases)
+    leases_to_remove = filter(inviolation_filter(termination_period), leases)
+
+    if delete:
+        for lease in leases_to_warn:
+            send_notification(
+                lease, sender, warn_period, termination_period,
+                "Your lease {} is idle and may be terminated.",
+                _email.IDLE_LEASE_WARNING_EMAIL_BODY)
+        for lease in leases_to_remove:
+            blazar.lease_delete(lease['id'])
+            send_notification(
+                lease, sender, warn_period, termination_period,
+                "Your lease {} has been terminated.",
+                _email.IDLE_LEASE_TERMINATION_EMAIL_BODY)
+
+    return leases_to_warn, leases_to_remove
+
+
+def main(argv=None):
+    if argv is None:
+        argv = sys.argv
+
+    parser = argparse.ArgumentParser(description='Unutilized Lease Reaper')
+    osapi.add_arguments(parser)
+    parser.add_argument(
+        '--slack', type=str,
+        help='JSON file with Slack webhook information')
+    parser.add_argument(
+        '-w', '--warn-hours', type=int,
+        help='Number of hours after which to warn user.',
+        default=DEFAULT_WARN_HOURS)
+    parser.add_argument(
+        '-r', '--grace-hours', type=int,
+        help='Number of hours after which to remove lease.',
+        default=DEFAULT_GRACE_HOURS)
+    parser.add_argument(
+        'action', choices=['info', 'delete'],
+        help='Just display info or actually delete them?')
+    parser.add_argument(
+        '--sender',
+        type=str,
+        help='Email address of sender.',
+        default='noreply@chameleoncloud.org')
+
+    args = parser.parse_args(argv[1:])
+    auth = osapi.Auth.from_env_or_args(args=args)
+
+    assert args.grace_hours > args.warn_hours, (
+        "Grace hours must be greater than warning period.")
+
+    if args.slack and (args.action == 'delete'):
```

Comment:
Feels weird to me that this won't run in delete mode if Slack is not configured, but I think other hammers do this as well. I would think, it would run the reaper, and if Slack is configured it will send the output to Slack, else it will print similar to how 'info' works.

Reply:
Yea, did it to be consistent. Also unclear why.